adam_core/observers/state.py

```diff
@@ -0,0 +1,147 @@
+    setup_SPICE()
+
+    # Get observatory geodetic information
+    geodetics = OBSERVATORY_GEODETICS.select("code", code).values[0]
+
+    if np.any(np.isnan(geodetics)):
+        err = (
+            f"{code} is missing information on Earth-based geodetic coordinates. The MPC Obs Code\n"
+            "file may be missing this information or the observer is a space-based observatory.\n"
+            "Space observatories are currently not supported.\n"
+        )
+        raise ValueError(err)
+
+    # Grab Earth state vector (this function handles duplicate times)
+    state = get_perturber_state(OriginCodes.EARTH, times, frame=frame, origin=origin)
+
+    # If the code is 500 (geocenter), we can just return the Earth state vector
+    if code == "500":
+        r_obs = state.r
+        v_obs = state.v
+
+    # If not then we need to add a topocentric correction
+    else:
+        # Get observer location on Earth
+        longitude = geodetics[0]
+        cos_phi = geodetics[1]
+        sin_phi = geodetics[2]
+        sin_longitude = np.sin(np.radians(longitude))
+        cos_longitude = np.cos(np.radians(longitude))
+
+        # Calculate pointing vector from geocenter to observatory
+        o_hat_ITRF93 = np.array(
+            [cos_longitude * cos_phi, sin_longitude * cos_phi, sin_phi]
+        )
+
+        # Multiply pointing vector with Earth radius to get actual vector
+        o_vec_ITRF93 = np.dot(R_EARTH, o_hat_ITRF93)
+
+        # Convert MJD epochs in TDB to ET in TDB
+        epochs_tdb = times.tdb.jd
+        unique_epochs_tdb = np.unique(epochs_tdb)
+        unique_epochs_et = np.array(
+            [sp.str2et("JD {:.16f} TDB".format(i)) for i in unique_epochs_tdb]
+        )
+
+        N = len(epochs_tdb)
+        r_obs = np.empty((N, 3), dtype=np.float64)
+        v_obs = np.empty((N, 3), dtype=np.float64)
+        r_geo = state.r
+        v_geo = state.v
+        for i, epoch in enumerate(unique_epochs_et):
+            # Grab rotaton matrices from ITRF93 to ecliptic J2000
+            # The ITRF93 high accuracy Earth rotation model takes into account:
+            # Precession:  1976 IAU model from Lieske.
+            # Nutation:  1980 IAU model, with IERS corrections due to Herring et al.
+            # True sidereal time using accurate values of TAI-UT1
+            # Polar motion
+            rotation_matrix = sp.pxform("ITRF93", frame_spice, epoch)
+
+            # Find indices of epochs that match the current unique epoch
+            mask = np.where(epochs_tdb == unique_epochs_tdb[i])[0]
+
+            # Add o_vec + r_geo to get r_obs (thank you numpy broadcasting)
+            r_obs[mask] = r_geo[mask] + rotation_matrix @ o_vec_ITRF93
+
+            # Calculate the velocity (thank you numpy broadcasting)
+            rotation_direction = np.cross(o_hat_ITRF93, Z_AXIS)
+            v_obs[mask] = v_geo[mask] + rotation_matrix @ (
+                -OMEGA_EARTH * R_EARTH * rotation_direction
+            )
+
```

Comment:
It seems like we're primarily iterating over the unique tdb representations, since we use that for generating the mask. I'm not sure if there's any performance gains from calculating the `unique_epochs_et` at the outset (I don't believe so, since it's using a simple list comprehension on generation and on fetching it we avoid the index lookup by generating it individually in the loop). 

```suggestion
        epochs_tdb = times.tdb.jd
        unique_epochs_tdb = np.unique(epochs_tdb)

        N = len(epochs_tdb)
        r_obs = np.empty((N, 3), dtype=np.float64)
        v_obs = np.empty((N, 3), dtype=np.float64)
        r_geo = state.r
        v_geo = state.v
        for epoch_tdb in unique_epochs_tdb:
            # Convert MJD epoch in TDB to ET in TDB
            epoch_et = sp.str2et("JD {:.16f} TDB".format(epoch_tbd)
            # Grab rotaton matrices from ITRF93 to ecliptic J2000
            # The ITRF93 high accuracy Earth rotation model takes into account:
            # Precession:  1976 IAU model from Lieske.
            # Nutation:  1980 IAU model, with IERS corrections due to Herring et al.
            # True sidereal time using accurate values of TAI-UT1
            # Polar motion

            rotation_matrix = sp.pxform("ITRF93", frame_spice, epoch_et)

            # Find indices of epochs that match the current unique epoch
            mask = np.where(epochs_tdb == epoch_tdb)[0]

            # Add o_vec + r_geo to get r_obs (thank you numpy broadcasting)
            r_obs[mask] = r_geo[mask] + rotation_matrix @ o_vec_ITRF93

            # Calculate the velocity (thank you numpy broadcasting)
            rotation_direction = np.cross(o_hat_ITRF93, Z_AXIS)
            v_obs[mask] = v_geo[mask] + rotation_matrix @ (
                -OMEGA_EARTH * R_EARTH * rotation_direction
            )

```